README.md
# Batchelor
A lovely little Node.js module to perform batch requests with the Google REST API. Simply does it.

## Google API Batch Requests
This is a project to solve a need for a missing feature in the wonderfully epic and useful [google/google-api-nodejs-client](https://github.com/google/google-api-nodejs-client). Currently, this cannot be used to post media and has not been tested with posting anything but JSON.

In theory this library could be used with other APIs, but has only been tested with Google's APIs as that's what we need it for.

Feel free to get involved in development.

## Upgrading to 2.0

`batch.run(...)` now takes a node style callback, with an error as the first parameter and the result as the second.

**INCORRECT** (old, v1.0-style):

``` node
batch.run(function(result){});
```

**CORRECT** (new, v2.0-style):

``` node
batch.run(function(err, result){});
```

## Upgrading to 1.0

The API was changed in 1.0 to move from a singleton instance to a constructor. So before where you used `Batchelor` directly:

``` node
var Batchelor = require('batchelor')
Batchelor.init(...)
Batchelor.add(...)
Batchelor.run(...)
```

You now need to create an instance of Batchelor:

``` node
var Batchelor = require('batchelor')
var batch = new Batchelor(...)
batch.add(...)
batch.run(...)
```

See <https://github.com/wapisasa/batchelor/issues/4> for why this change was made.

## Installation

This library has also been distributed on `npm`. Install it with the following command:

``` sh
$ npm install batchelor --save
```

## How to Use
#### GET Requests
``` node
var Batchelor = require('batchelor');
```
Once the module has been included, we initialise it with all our default options:
``` node
var batch = new Batchelor({
	// Any batch uri endpoint in the form: https://www.googleapis.com/batch/<api>/<version>
	'uri':'https://www.googleapis.com/batch/gmail/v1/',
	'method':'POST',
	'auth': {
		'bearer': [... Google API Token ...]
	},
	'headers': {
		'Content-Type': 'multipart/mixed'
	}
});
```
We can then start adding requests to our batch. This can be done 2 ways:

As a one-off object:
``` node
batch.add({
	'method':'GET',
	'path':'/gmail/v1/users/me/messages/123'
})
```
Or an Array of objects:
``` node
batch.add([
	{
		'method':'GET',
		'path':'/gmail/v1/users/me/messages/123'
	},
	{
		'method':'GET',
		'path':'/gmail/v1/users/me/messages/456'
	},
	{
		'method':'GET',
		'path':/gmail/v1/users/me/messages/789'
	}
]);
```
Once you have added all of the requests you need, call `.run()`:
``` node
batch.run(function(err, response){
	if (err){
		console.log("Error: " + err.toString());
	} else {
		res.json(response);
	}
});
```
#### POST Requests
The above examples show `GET` requests. To perform a `POST` requires a few more settings:
``` node
batch.add({
	'method':'POST',
	'path':'/gmail/v1/users/me/messages/',
	'parameters':{
		'Content-Type':'application/json;',
		'body':{'object':{'originalContent': 'A wonderful batch post!'},'access': {'items': [{'type': 'domain'}],'domainRestricted': true}}
	}
});
```
#### Callbacks
By default, all responses are returned through the callback function in the `batch.run()` call. Alternatively, a callback can be supplied for each individual calls:
``` node
batch.add({
	'method':'POST',
	'path':'/gmail/v1/users/me/messages/',
	'parameters':{
		'Content-Type':'application/json;',
		'body':{'object':{'originalContent': 'Another wonderful batch post with callback!'},'access': {'items': [{'type': 'domain'}],'domainRestricted': true}}
	},
	'callback':function(response){
		console.log(response);
	}
});
```
#### Request and Response IDs
The module will assign a request a randomly generated unique `Content-ID` by default, but this can be supplied as part of the options to supply `batch.add()`:
``` node
batch.add({
	'method':'GET',
	'path':'/gmail/v1/users/me/messages/123',
	'requestId':'Batch_UniqueID_1'
})
```
#### A Couple of Little Gifts
###### Method Chaining
All methods return the `Batchelor` instance. So you can chain calls together.

``` node
batch.add([
	...
]).run(function(err, data){
	...
});
```
###### Data Pass-through
When passing options to the `.add()` you can include an Object called `extend`. In the case of providing a callback, this will be passed back as a second parameter. When using the default callback on the `.run()` call, an array of all data passed through will be added as a second parameter with the requestId as the key:
``` node
batch.add({
	...
	'extend':{
		...
	},
	'callback':function(response, extendObj){
		console.log(response, extendObj);
	}
});
```
This could be required, for example, when making multiple requests with different Auth data and then needing to make further requests with the same Auth data.

###### Resetting and Re-using
Once Batchelor has been run, there are certain use-cases for running futher batch requests on response. This requires the variables in the instance to be reset. This can be done using the `.reset()` call:
``` node
batch.run(function(err, response){

	//	Reset Batchelor for further use
	batch.reset();
	...

});
```

## To Do List-ish
These might get done if we end up needing them/have time:
* Limit requests per batch request
* Handle Media in API calls (no need for it here, feel free to write it)

## Building

`npm run build`

Outputs a commonjs-compatible bundle to `dist/index.js`.

## Running tests

`npm test`

## Publishing

Merging to master will automatically publish the package if commits with non-trivial changes have
been introduced (per [commit conventions](https://www.conventionalcommits.org)).

## Acknowledgement
Built on the clock by [@jamesmhaley](https://github.com/jamesmhaley) as an internal project of [wapisasa C.I.C.](http://www.wapisasa.com)


CHANGELOG.md

* 2.0.2 Added [error handler](https://github.com/wapisasa/batchelor/pull/15) and [made `headers` optional](https://github.com/wapisasa/batchelor/pull/17)
* 2.0.1 Include response status code in error
* 2.0.0 Bug Fixes and Improvements
* 1.0.0 The API was changed in 1.0 to move from a singleton instance to a constructor (Thanks again to [@bradvogel](https://github.com/bradvogel))
* 0.0.8 Added support for DELETE requests (Thanks to [@bradvogel](https://github.com/bradvogel))
* 0.0.7 Reset function for when you are using batchelor more than once in a script (ability for nested requests too)
* 0.0.6 Bug fixes introduced in the last update and clean up was happening too soon. Moved it.
* 0.0.5 Added the ability to passthrough specific information from the `.add()` options to the response object
* 0.0.4 Authorization can now be set on each request (currently Bearer [...] only)
* 0.0.3 More bug fixes: New contentId for every request
* 0.0.2 Bug fixes
* 0.0.1 Initial release


src/index.test.js
import Batchelor from './';
import nock from 'nock';
import util from 'util';
import fs from 'fs';
import expectedResult from './mocks/result.json';

beforeAll(() => {
  nock.disableNetConnect();
});

describe('Batchelor', () => {
  let client;

  beforeEach(() => {
    client = new Batchelor({
      uri: 'https://www.googleapis.com/batch/gmail/v1/',
      method: 'POST',
      auth: {
        bearer: '',
      },
      headers: {
        'Content-Type': 'multipart/mixed',
      },
    });
    // Make promise based.
    client.run = util.promisify(client.run);
  });

  it('should request an end-point', async () => {
    const scope = nock('https://www.googleapis.com')
      .post('/batch/gmail/v1/')
      .replyWithFile(200, __dirname + '/mocks/response.txt', {
        'Content-Type': 'multipart/mixed; boundary=batch_e18a79KIjgaVUvAf-cfwbL1MtYtcjIqC',
      });

    client.add({
      method: 'GET',
      path: '/gmail/v1/users/me/messages/172166a90d50fe37',
    });
    client.add({
      method: 'GET',
      path: '/gmail/v1/users/me/messages/172163f16b503c65',
    });
    const result = await client.run();

    expect(result).toEqual(expectedResult);
    scope.done();
  });
});


src/index.js
/*
  The MIT License (MIT)

  Copyright (c)2015 wapisasa C.I.C.

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in all
  copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  SOFTWARE.
*/

'use strict';

var request = require('request'),
  Dicer = require('dicer'),
  parser = require('http-string-parser'),
  hat = require('hat');


/**
 * Initialise Batchelor
 *
 * @options {object} options for the parent request
 * @return  {object} Batchelor object
 */
var Batchelor = function (options) {

  //  URI is only required option
  if (!options.uri) {
    throw new Error('options.uri is required');
  }

  //  Defaults options
  options.method = options.method || 'POST'; // Default is POST as this was built specifically for Google
  if (!options.headers) options.headers = {};
  options.headers['Content-Type'] = options.headers['Content-Type'] || 'multipart/mixed;';

  //  Globalise the options
  this._globalOptions = options;

  //  Ready internal variables
  this._requests = [];
  this._requestSpecificCallbacks = [];
  this._requestExtensionData = [];
};

/**
 * Add a request into the requests array
 *
 * @options {object/array} options for individual requests/array of requests
 * @return  {object} Batchelor object
 */
Batchelor.prototype.add = function (options) {

  var _self = this;

  //  Check if adding multiple requests and loop through array
  if (!!Array.isArray(options)) {
    options.forEach(_self.add, _self);

    return _self;
  }

  //  Check for required options
  if (!options.path) {
    throw new Error('options.path is a required argument');
  }

  //  Give each request an id so we can identify the response
  var rack = hat.rack(),
    requestId = options.requestId || 'Batchelor_' + rack();

  //  Save out request specific callback
  if (!!options.callback) {
    this._requestSpecificCallbacks[requestId] = options.callback;
  }

  //  Any data thay needs to be passed through to the callback
  if (!!options.extend) {
    this._requestExtensionData[requestId] = options.extend;
  }

  //  Set Defaults
  options.method = options.method || 'GET';
  options.requestId = requestId;
  if (options.method === 'POST') {
    //  It's not a GET so we need something to push
    if (!options.parameters) {
      throw new Error('when using POST: options.parameters is required');
    }

    //  We will asume if it's not set, we are sending a body of JSON
    options.parameters['Content-Type'] = options.parameters['Content-Type'] || 'application/json;';

    //  Body exists?
    if (!options.parameters.body) {
      throw new Error('when using POST: options.parameters.body is required');
    }
  }

  //  Push to _requests Array
  this._requests.push(options);

  //  Chaining baby
  return this;

};

/**
 * Run the batch requests
 *
 * @callback {function} function to run when request is complete
 * @return   {object} API response via callback
 */
Batchelor.prototype.run = function (callback) {

  var _multiparts = [],
    _self = this;

  //  Build multipart request
  this._requests.forEach(function (requestPart) {

    var requestSettings = {
      'Content-Type': 'application/http',
      'Content-ID': requestPart.requestId,
      'body': requestPart.method + ' ' + requestPart.path + ' HTTP/1.1\n'
    };

    //  Check if this part needs it's own auth
    if ((!!requestPart.auth) && (!!requestPart.auth.bearer)) {
      requestSettings.body += 'Authorization: Bearer ' + requestPart.auth.bearer + '\n';
    }

    //  Replace body with a request if this isn't a GET
    if (requestPart.method !== 'GET') {
      requestSettings.body = requestPart.method + ' ' + requestPart.path + ' HTTP/1.1\n' +
        'Content-Type: ' + requestPart.parameters['Content-Type'] + '\n\n' +
        JSON.stringify(requestPart.parameters.body);
    }

    //  Push into array for the request
    _multiparts.push(requestSettings);

  });

  //  Collate objects
  var requestObj = this._globalOptions;
  requestObj.multipart = _multiparts;

  //  Run Requests
  request(requestObj).on('response', function (responseObj) {

    //  TODO: Handle errors bette than this
    if (responseObj.statusCode !== 200) {
      callback(new Error('Request did not return 200 OK: ' + responseObj.statusCode));
      return;
    }

    //  Get the boundary
    var boundaryRegex = /^multipart\/.+?(?:; boundary=(?:(?:"(.+)")|(?:([^\s]+))))$/i,
      boundaries = boundaryRegex.exec(responseObj.headers['content-type']),
      boundary = boundaries[1] || boundaries[2],
      responseBody = {
        'parts': [],
        'errors': 0
      };

    //  Cannot find our boundary
    if (!boundary) {
      callback(new Error('Problem getting boundary data'));
      return;
    }

    //  Now we read our body and output
    var d = new Dicer({
      boundary: boundary
    });

    //  Roll through our parts and build the bigger picture
    d.on('part', function (p) {

      var part = {
        data: [],
        bodylen: 0,
        header: undefined
      };

      p.on('header', function (h) {

        //  Save header data
        part.header = h;

      }).on('data', function (data) {

        //  Build this part's body
        part.data.push(data);
        part.bodylen += data.length;

      }).on('end', function () {

        //  Bring body parts together
        if (part.data) {
          part.data = Buffer.concat(part.data, part.bodylen);
        }

        //  Parse the raw http response
        part.data = parser.parseResponse(part.data.toString());

        //  Parse response JSON
        if (part.data.body) {
          part.data.body = JSON.parse(part.data.body);
        }

        //  Get the response id if exists
        var returnedContentId = part.header['content-id'],
          currentRequestCallback,
          currentRequestExtendData;

        if (!!returnedContentId) {

          //  In some cases (Google API) this is a single item array, flatten it and remove "response" identifier.
          if (!!Array.isArray(returnedContentId) && returnedContentId.length < 2) {
            returnedContentId = returnedContentId.join('').replace(/response-/, '');
          }

          part.data.headers['Content-ID'] = returnedContentId;

          //  As we have a content id, we can link to individual callback (if exists)
          currentRequestCallback = _self._requestSpecificCallbacks[returnedContentId];

          //  If this request has extend data, retrieve it for callback
          currentRequestExtendData = _self._requestExtensionData[returnedContentId];
        }

        //  If this individual request has a callback, run it
        if (!!currentRequestCallback) {

          currentRequestCallback(part.data, currentRequestExtendData || null);

        } else {
          //  Anything that doesn't have a specific callback get's pushed into main body
          responseBody.parts.push(part.data);
        }

      });
    }).on('finish', function () {

      //  All remaining responses get output
      //  Extend data sent through as is for parse in callback
      callback(null /* No error */ , responseBody, _self._requestExtensionData);

    });

    //  Pipe our response to Dicer
    responseObj.pipe(d);
  }).on('error', function (err) {
    // report socket errors to the callback function
    callback(err);
  });
};

/**
 * Reset all internal options if a re-run is required
 *
 * @return  {object} Batchelor object
 */
Batchelor.prototype.reset = function () {

  this._requests = [];
  this._requestSpecificCallbacks = [];
  this._requestExtensionData = [];
  this._globalOptions.multipart = [];

  return this;
};

module.exports = Batchelor;


lib/batchelor.js
/*
  The MIT License (MIT)

  Copyright (c)2015 wapisasa C.I.C.

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in all
  copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  SOFTWARE.
*/

'use strict';

var request = require('request'),
  Dicer = require('dicer'),
  parser = require('http-string-parser'),
  hat = require('hat');


/**
 * Initialise Batchelor
 *
 * @options {object} options for the parent request
 * @return  {object} Batchelor object
 */
var Batchelor = function (options) {

  //  URI is only required option
  if (!options.uri) {
    throw new Error('options.uri is required');
  }

  //  Defaults options
  options.method = options.method || 'POST'; // Default is POST as this was built specifically for Google
  if (!options.headers) options.headers = {};
  options.headers['Content-Type'] = options.headers['Content-Type'] || 'multipart/mixed;';

  //  Globalise the options
  this._globalOptions = options;

  //  Ready internal variables
  this._requests = [];
  this._requestSpecificCallbacks = [];
  this._requestExtensionData = [];
};

/**
 * Add a request into the requests array
 *
 * @options {object/array} options for individual requests/array of requests
 * @return  {object} Batchelor object
 */
Batchelor.prototype.add = function (options) {

  var _self = this;

  //  Check if adding multiple requests and loop through array
  if (!!Array.isArray(options)) {
    options.forEach(_self.add, _self);

    return _self;
  }

  //  Check for required options
  if (!options.path) {
    throw new Error('options.path is a required argument');
  }

  //  Give each request an id so we can identify the response
  var rack = hat.rack(),
    requestId = options.requestId || 'Batchelor_' + rack();

  //  Save out request specific callback
  if (!!options.callback) {
    this._requestSpecificCallbacks[requestId] = options.callback;
  }

  //  Any data thay needs to be passed through to the callback
  if (!!options.extend) {
    this._requestExtensionData[requestId] = options.extend;
  }

  //  Set Defaults
  options.method = options.method || 'GET';
  options.requestId = requestId;
  if (options.method === 'POST') {
    //  It's not a GET so we need something to push
    if (!options.parameters) {
      throw new Error('when using POST: options.parameters is required');
    }

    //  We will asume if it's not set, we are sending a body of JSON
    options.parameters['Content-Type'] = options.parameters['Content-Type'] || 'application/json;';

    //  Body exists?
    if (!options.parameters.body) {
      throw new Error('when using POST: options.parameters.body is required');
    }
  }

  //  Push to _requests Array
  this._requests.push(options);

  //  Chaining baby
  return this;

};

/**
 * Run the batch requests
 *
 * @callback {function} function to run when request is complete
 * @return   {object} API response via callback
 */
Batchelor.prototype.run = function (callback) {

  var _multiparts = [],
    _self = this;

  //  Build multipart request
  this._requests.forEach(function (requestPart) {

    var requestSettings = {
      'Content-Type': 'application/http',
      'Content-ID': requestPart.requestId,
      'body': requestPart.method + ' ' + requestPart.path + ' HTTP/1.1\n'
    };

    //  Check if this part needs it's own auth
    if ((!!requestPart.auth) && (!!requestPart.auth.bearer)) {
      requestSettings.body += 'Authorization: Bearer ' + requestPart.auth.bearer + '\n';
    }

    //  Replace body with a request if this isn't a GET
    if (requestPart.method !== 'GET') {
      requestSettings.body = requestPart.method + ' ' + requestPart.path + ' HTTP/1.1\n' +
        'Content-Type: ' + requestPart.parameters['Content-Type'] + '\n\n' +
        JSON.stringify(requestPart.parameters.body);
    }

    //  Push into array for the request
    _multiparts.push(requestSettings);

  });

  //  Collate objects
  var requestObj = this._globalOptions;
  requestObj.multipart = _multiparts;

  //  Run Requests
  request(requestObj).on('response', function (responseObj) {

    //  TODO: Handle errors bette than this
    if (responseObj.statusCode !== 200) {
      callback(new Error('Request did not return 200 OK: ' + responseObj.statusCode));
      return;
    }

    //  Get the boundary
    var boundaryRegex = /^multipart\/.+?(?:; boundary=(?:(?:"(.+)")|(?:([^\s]+))))$/i,
      boundaries = boundaryRegex.exec(responseObj.headers['content-type']),
      boundary = boundaries[1] || boundaries[2],
      responseBody = {
        'parts': [],
        'errors': 0
      };

    //  Cannot find our boundary
    if (!boundary) {
      callback(new Error('Problem getting boundary data'));
      return;
    }

    //  Now we read our body and output
    var d = new Dicer({
      boundary: boundary
    });

    //  Roll through our parts and build the bigger picture
    d.on('part', function (p) {

      var part = {
        data: [],
        bodylen: 0,
        header: undefined
      };

      p.on('header', function (h) {

        //  Save header data
        part.header = h;

      }).on('data', function (data) {

        //  Build this part's body
        part.data.push(data);
        part.bodylen += data.length;

      }).on('end', function () {

        //  Bring body parts together
        if (part.data) {
          part.data = Buffer.concat(part.data, part.bodylen);
        }

        //  Parse the raw http response
        part.data = parser.parseResponse(part.data.toString());

        //  Parse response JSON
        if (part.data.body) {
          try {
            part.data.body = JSON.parse(part.data.body);
          } catch (e) {
            part.data.body = {parseError: part.data.body};
          }
        }

        //  Get the response id if exists
        var returnedContentId = part.header['content-id'],
          currentRequestCallback,
          currentRequestExtendData;

        if (!!returnedContentId) {

          //  In some cases (Google API) this is a single item array, flatten it and remove "response" identifier.
          if (!!Array.isArray(returnedContentId) && returnedContentId.length < 2) {
            returnedContentId = returnedContentId.join('').replace(/response-/, '');
          }

          part.data.headers['Content-ID'] = returnedContentId;

          //  As we have a content id, we can link to individual callback (if exists)
          currentRequestCallback = _self._requestSpecificCallbacks[returnedContentId];

          //  If this request has extend data, retrieve it for callback
          currentRequestExtendData = _self._requestExtensionData[returnedContentId];
        }

        //  If this individual request has a callback, run it
        if (!!currentRequestCallback) {

          currentRequestCallback(part.data, currentRequestExtendData || null);

        } else {
          //  Anything that doesn't have a specific callback get's pushed into main body
          responseBody.parts.push(part.data);
        }

      });
    }).on('finish', function () {

      //  All remaining responses get output
      //  Extend data sent through as is for parse in callback
      callback(null /* No error */ , responseBody, _self._requestExtensionData);

    });

    //  Pipe our response to Dicer
    responseObj.pipe(d);
  }).on('error', function (err) {
    // report socket errors to the callback function
    callback(err);
  });
};

/**
 * Reset all internal options if a re-run is required
 *
 * @return  {object} Batchelor object
 */
Batchelor.prototype.reset = function () {

  this._requests = [];
  this._requestSpecificCallbacks = [];
  this._requestExtensionData = [];
  this._globalOptions.multipart = [];

  return this;
};

module.exports = Batchelor;
